File: Reorder Routes to Make All Paths Lead to the City Zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    public void dfs(int node, List<List<Integer>> away, List<List<Integer>> tow, 
                    int[] ans, boolean[] vis) {
        vis[node] = true;
        for (int i : away.get(node)) {
            if (!vis[i]) {
                ans[0]++;
                dfs(i, away, tow, ans, vis);
            }
        }
        for (int i : tow.get(node)) {
            if (!vis[i]) {
                dfs(i, away, tow, ans, vis);
            }
        }
    }

    public int minReorder(int n, int[][] connections) {
        List<List<Integer>> away = new ArrayList<>();
        List<List<Integer>> tow = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            away.add(new ArrayList<>());
            tow.add(new ArrayList<>());
        }

        for (int[] connection : connections) {
            int a = connection[0];
            int b = connection[1];
            away.get(a).add(b);
            tow.get(b).add(a);
        }

        boolean[] vis = new boolean[n];
        int[] ans = {0};

        
        dfs(0, away, tow, ans, vis);

        return ans[0];
    }
}

File: Reorder Routes to Make All Paths Lead to the City Zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.cpp
class Solution {
public:
    void dfs(int node, vector<vector<int>>& away, vector<vector<int>>& tow,
             int& ans, vector<int>& vis) {
        vis[node] = 1;
        for (int i : away[node]) {
            if (!vis[i]) {
                ans += 1;
                dfs(i, away, tow, ans, vis);
            }
        }
        for (int i : tow[node]) {
            if (!vis[i]) {
                dfs(i, away, tow, ans, vis);
            }
        }
    }
    int minReorder(int n, vector<vector<int>>& connections) {
        vector<vector<int>> away(n), tow(n);
        vector<int> vis(n, 0);
        for (int i = 0; i < connections.size(); i++) {
            int a = connections[i][0];
            int b = connections[i][1];
            away[a].push_back(b);
            tow[b].push_back(a);
        }
        int ans = 0;
        dfs(0, away, tow, ans, vis);
        return ans;
    }
};